Can not be executed on Windows 10

OS: Windows 10 Pro

Since I'm not a big fan of -g installs, I've created an empty npm project, and installed appium-doctor using
`npm install --save appium appium-doctor`

Then by adding appium-doctor in scripts section of package.json I'm able to execute.

But receiving this error:
```
TypeError: Cannot read property 'replace' of undefined
    at Function.retrieveUsingWhichCommand$ (X:\ondra\internal\appium-test\node_modules\appium-doctor\build\lib\node-detector.js:152:30)
    at tryCatch (X:\ondra\internal\appium-test\node_modules\babel-runtime\regenerator\runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (X:\ondra\internal\appium-test\node_modules\babel-runtime\regenerator\runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (X:\ondra\internal\appium-test\node_modules\babel-runtime\regenerator\runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (X:\ondra\internal\appium-test\node_modules\babel-runtime\regenerator\runtime.js:136:37)
    at process._tickCallback (internal/process/next_tick.js:103:7)

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "appium-doctor"
npm ERR! node v6.9.2
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! appium-test@1.0.0 appium-doctor: `appium-doctor`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the appium-test@1.0.0 appium-doctor script 'appium-doctor'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the appium-test package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     appium-doctor
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs appium-test
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls appium-test
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     \appium-test\npm-debug.log
```

Looking in the code I've noticed you can set NODE_BIN variable to skip all the automatic lookups for a node. It might be stated in the readme, as recommended on Windows.

There is also a one-liner to set the variable using Windows' `where`:
`for /f "delims==" %i in ('where node') do @set "NODE_BIN=%i"`

--- a/lib/node-detector.js
+++ b/lib/node-detector.js
@@ -47,6 +47,11 @@ class NodeDetector {
   }
 
   static async retrieveUsingAppleScript () {
+    if (!system.isMac()) {
+      log.debug('Not on Darwin, skipping Apple Script');
+      return null;
+    }
+
     var appScript = [
       'try'
       , '  set appiumIsRunning to false'

--- a/lib/node-detector.js
+++ b/lib/node-detector.js
@@ -1,4 +1,4 @@
-import { fs } from 'appium-support';
+import { fs, system } from 'appium-support';
 import { exec } from 'teen_process';
 import log from './logger';
 import path from 'path';
@@ -22,38 +22,26 @@ class NodeDetector {
     return null;
   }
 
-  static async retrieveUsingWhichCommand () {
+  static async retrieveUsingSystemCall () {
     let stdout;
-    try {
-      stdout = (await exec('which', ['node'])).stdout;
-    } catch (err) {
-      log.debug(err);
-      return null;
-    }
-    let nodePath = stdout.replace("\n", "");
-    if (await fs.exists(nodePath)) {
-      log.debug(`Node binary found using which command at: ${nodePath}`);
-      return nodePath;
-    } else {
-      log.debug('Node binary not found using the which command.');
-      return null;
-    }
-  }
+    let cmd = 'which';
 
-  static async retrieveUsingWhereCommand () {
-    let stdout;
+    if (system.isWindows()) {
+      cmd = 'where';
+    }
+
     try {
-      stdout = (await exec('where', ['node'])).stdout;
+      stdout = (await exec(cmd, ['node'])).stdout;
     } catch (err) {
       log.debug(err);
       return null;
     }
-    let nodePath = stdout.replace("\n", "");
+    let nodePath = stdout.replace(/[\n\r]/g, "");
     if (await fs.exists(nodePath)) {
-      log.debug(`Node binary found using where command at: ${nodePath}`);
+      log.debug(`Node binary found using ${cmd} command at: ${nodePath}`);
       return nodePath;
     } else {
-      log.debug('Node binary not found using the where command.');
+      log.debug(`Node binary not found using the ${cmd} command.`);
       return null;
     }
   }
@@ -110,8 +98,7 @@ class NodeDetector {
 
   static async detect () {
     let nodePath = await this.retrieveInCommonPlaces() ||
-      await NodeDetector.retrieveUsingWhichCommand() ||
-      await NodeDetector.retrieveUsingWhereCommand() ||
+      await NodeDetector.retrieveUsingSystemCall() ||
       await NodeDetector.retrieveUsingAppleScript() ||
       await NodeDetector.retrieveUsingAppiumConfigFile();
     if (nodePath) {

--- a/lib/node-detector.js
+++ b/lib/node-detector.js
@@ -26,7 +26,10 @@ class NodeDetector {
     let stdout;
     try {
       stdout = (await exec('which', ['node'])).stdout;
-    } catch (ign) {}
+    } catch (ign) {
+      log.debug(ign);
+      return null;
+    }
     let nodePath = stdout.replace("\n", "");
     if (await fs.exists(nodePath)) {
       log.debug(`Node binary found using which command at: ${nodePath}`);
@@ -53,7 +56,10 @@ class NodeDetector {
     let stdout;
     try {
       stdout = (await exec('osascript', ['-e', appScript])).stdout;
-    } catch (ign) {}
+    } catch (ign) {
+        log.debug(ign);
+        return null;
+    }
     let nodePath = stdout.replace("\n", "");
     if (await fs.exists(nodePath)) {
       log.debug(`Node binary found using AppleScript at: ${nodePath}`);
@@ -71,7 +77,10 @@ class NodeDetector {
       if (await fs.exists(appiumConfigPath)) {
         jsonobj = JSON.parse(await fs.readFile(appiumConfigPath, 'utf8'));
       }
-    } catch (ign) {}
+    } catch (ign) {
+        log.debug(ign);
+        return null;
+    }
     if (jsonobj && jsonobj.node_bin && await fs.exists(jsonobj.node_bin) ) {
       log.debug(`Node binary found using .appiumconfig.json at: ${jsonobj.node_bin}`);
       return jsonobj.node_bin;

--- a/lib/node-detector.js
+++ b/lib/node-detector.js
@@ -26,8 +26,8 @@ class NodeDetector {
     let stdout;
     try {
       stdout = (await exec('which', ['node'])).stdout;
-    } catch (ign) {
-      log.debug(ign);
+    } catch (err) {
+      log.debug(err);
       return null;
     }
     let nodePath = stdout.replace("\n", "");
@@ -44,8 +44,8 @@ class NodeDetector {
     let stdout;
     try {
       stdout = (await exec('where', ['node'])).stdout;
-    } catch (ign) {
-      log.debug(ign);
+    } catch (err) {
+      log.debug(err);
       return null;
     }
     let nodePath = stdout.replace("\n", "");
@@ -74,8 +74,8 @@ class NodeDetector {
     let stdout;
     try {
       stdout = (await exec('osascript', ['-e', appScript])).stdout;
-    } catch (ign) {
-      log.debug(ign);
+    } catch (err) {
+      log.debug(err);
       return null;
     }
     let nodePath = stdout.replace("\n", "");
@@ -95,8 +95,8 @@ class NodeDetector {
       if (await fs.exists(appiumConfigPath)) {
         jsonobj = JSON.parse(await fs.readFile(appiumConfigPath, 'utf8'));
       }
-    } catch (ign) {
-      log.debug(ign);
+    } catch (err) {
+      log.debug(err);
       return null;
     }
     if (jsonobj && jsonobj.node_bin && await fs.exists(jsonobj.node_bin) ) {

--- a/lib/node-detector.js
+++ b/lib/node-detector.js
@@ -40,6 +40,24 @@ class NodeDetector {
     }
   }
 
+  static async retrieveUsingWhereCommand () {
+    let stdout;
+    try {
+      stdout = (await exec('where', ['node'])).stdout;
+    } catch (ign) {
+      log.debug(ign);
+      return null;
+    }
+    let nodePath = stdout.replace("\n", "");
+    if (await fs.exists(nodePath)) {
+      log.debug(`Node binary found using where command at: ${nodePath}`);
+      return nodePath;
+    } else {
+      log.debug('Node binary not found using the where command.');
+      return null;
+    }
+  }
+
   static async retrieveUsingAppleScript () {
     var appScript = [
       'try'
@@ -57,8 +75,8 @@ class NodeDetector {
     try {
       stdout = (await exec('osascript', ['-e', appScript])).stdout;
     } catch (ign) {
-        log.debug(ign);
-        return null;
+      log.debug(ign);
+      return null;
     }
     let nodePath = stdout.replace("\n", "");
     if (await fs.exists(nodePath)) {
@@ -78,8 +96,8 @@ class NodeDetector {
         jsonobj = JSON.parse(await fs.readFile(appiumConfigPath, 'utf8'));
       }
     } catch (ign) {
-        log.debug(ign);
-        return null;
+      log.debug(ign);
+      return null;
     }
     if (jsonobj && jsonobj.node_bin && await fs.exists(jsonobj.node_bin) ) {
       log.debug(`Node binary found using .appiumconfig.json at: ${jsonobj.node_bin}`);
@@ -93,6 +111,7 @@ class NodeDetector {
   static async detect () {
     let nodePath = await this.retrieveInCommonPlaces() ||
       await NodeDetector.retrieveUsingWhichCommand() ||
+      await NodeDetector.retrieveUsingWhereCommand() ||
       await NodeDetector.retrieveUsingAppleScript() ||
       await NodeDetector.retrieveUsingAppiumConfigFile();
     if (nodePath) {

--- a/test/node-detector-specs.js
+++ b/test/node-detector-specs.js
@@ -1,7 +1,7 @@
 // transpile:mocha
 
 import chai from 'chai';
-import { fs } from 'appium-support';
+import { fs, system } from 'appium-support';
 import * as tp from 'teen_process';
 import NodeDetector from '../lib/node-detector';
 import B from 'bluebird';
@@ -25,8 +25,11 @@ describe('NodeDetector', withMocks({fs, tp}, (mocks, S) => {
     verify(mocks);
   });
 
-  // retrieveUsingWhichCommand
+  // retrieveUsingSystemCall
   let testRetrieveWithScript = (method) => {
+    if (method === 'retrieveUsingAppleScript') {
+      system.isMac = () => true;
+    }
     it(method + ' - success', async () => {
       mocks.tp.expects('exec').once().returns(
         B.resolve({stdout: '/a/b/c/d\n', stderr: ''}));
@@ -50,7 +53,7 @@ describe('NodeDetector', withMocks({fs, tp}, (mocks, S) => {
     });
   };
 
-  testRetrieveWithScript('retrieveUsingWhichCommand');
+  testRetrieveWithScript('retrieveUsingSystemCall');
   testRetrieveWithScript('retrieveUsingAppleScript');
 
   it('retrieveUsingAppiumConfigFile - success', async () => {
@@ -84,7 +87,7 @@ describe('NodeDetector', withMocks({fs, tp}, (mocks, S) => {
   it('checkForNodeBinary - success', async () => {
     mocks.NodeDetector = S.sandbox.mock(NodeDetector);
     mocks.NodeDetector.expects('retrieveInCommonPlaces').once().returns(null);
-    mocks.NodeDetector.expects('retrieveUsingWhichCommand').once().returns(null);
+    mocks.NodeDetector.expects('retrieveUsingSystemCall').once().returns(null);
     mocks.NodeDetector.expects('retrieveUsingAppleScript').returns('/a/b/c/d');
     mocks.NodeDetector.expects('retrieveUsingAppiumConfigFile').never();
     (await NodeDetector.detect()).should.equal('/a/b/c/d');
@@ -94,7 +97,7 @@ describe('NodeDetector', withMocks({fs, tp}, (mocks, S) => {
   it('checkForNodeBinary - failure', async () => {
     mocks.NodeDetector = S.sandbox.mock(NodeDetector);
     mocks.NodeDetector.expects('retrieveInCommonPlaces').once().returns(null);
-    mocks.NodeDetector.expects('retrieveUsingWhichCommand').once().returns(null);
+    mocks.NodeDetector.expects('retrieveUsingSystemCall').once().returns(null);
     mocks.NodeDetector.expects('retrieveUsingAppleScript').once().returns(null);
     mocks.NodeDetector.expects('retrieveUsingAppiumConfigFile').once().returns(null);
     expect(await NodeDetector.detect()).to.be.a('null');

